compose: Add "New topic" button to topic field in compose box.

shameondev opened the pull request:
This pull request introduces two changes:
1. It expands the width of the topic input to occupy the entire available space.
2. It adds a "New topic" button inside the input (visually). To achieve this, the following modifications were made:
    - Instead of a single input element, a div container was created to hold both the input and the button.
    - The tests and selectors were adjusted to correctly show and hide the new button when switching between DM and stream messages.
    - Some additional logic was implemented to handle the functionality of the "new topic" button.
    - The styles around the custom button container were modified.
    - JavaScript code was added to apply the "focused" class to the button container, ensuring that it displays the associated styles when the input inside the container receives focus.

Fixes: #24889.

[Feedback topic in CZO](https://chat.zulip.org/#narrow/stream/137-feedback/topic/new.20topic.20button)
[Issues topic in CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/new.20topic.20button)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Screen captures:</summary>

![2023-06-14 at 20 15 06 - Emerald Raccoon](https://github.com/zulip/zulip/assets/53193850/b12f3a78-0279-46b2-a51f-692478ca5d62)

</details>

<details>
<summary>Screenshots</summary>

| Light | Dark |
| ------ | ------- |
| No focus |
| ![image](https://github.com/zulip/zulip/assets/53193850/a448c527-da30-4c20-86f4-151fd12db67f) | ![image](https://github.com/zulip/zulip/assets/53193850/7115ba39-ae37-468e-9c56-0c011cf8f6ba) |
| Focused |
| ![image](https://github.com/zulip/zulip/assets/53193850/41924360-8bf5-4aed-8219-b7a551f91b17) | ![image](https://github.com/zulip/zulip/assets/53193850/d0de82e9-4d96-4b2f-bf35-f982b7083fd2)  |
| No focus |
| ![image](https://github.com/zulip/zulip/assets/53193850/82f1d17f-9e2b-4d0b-bba9-74c766223db6) | ![image](https://github.com/zulip/zulip/assets/53193850/6f9a6641-236e-406c-92f5-7f002b559b2e) |
| Focused |
| ![image](https://github.com/zulip/zulip/assets/53193850/9d399a3b-32e4-4ced-85a2-53753f6e7564) | ![image](https://github.com/zulip/zulip/assets/53193850/da8e5855-9820-4431-a4d3-c653e5eae259) |
 
</details>

karlstolley commented on web/templates/compose.hbs:
```diff
@@ -70,7 +70,10 @@
                                 <i class="fa fa-angle-right" aria-hidden="true"></i>
                             </div>
-                            <input type="text" class="recipient_box" name="stream_message_recipient_topic" id="stream_message_recipient_topic" maxlength="{{ max_topic_length }}" value="" placeholder="{{t 'Topic' }}" autocomplete="off" tabindex="0" aria-label="{{t 'Topic' }}" />
+                            <div id="compose_recipient_box">
```
This is model HTML for how to handle these kinds of "illusion" inputs, where one interactive element (a button) appears to sit inside another (an input). Nice work!

karlstolley commented on web/styles/compose.css:
```diff
@@ -535,11 +535,49 @@ textarea.new_message_textarea,
+    #recipient_box_new_topic_button {
+        flex-shrink: 0;
+        /* The button looks better if it have smaller radius than the parent */
+        border-radius: 2px;
```
These are small values, but did you eyeball that value? It's pretty easy to calculate harmonious border-radius values for nested elements: https://www.30secondsofcode.org/css/s/nested-border-radius/

shameondev replied:
Thanks for sharing that, it's really interesting. Here's the quote from the post you provided:

> This can be expressed mathematically as innerRadius + distance = outerRadius, or more simply R1 + D = R2.

 I did exactly the same thing, just without the theoretical background. In our case, we have 3 pixels on the parent, 1 pixel for the margin, and the resulting value for the inner radius is 2 pixels.

karlstolley commented on web/styles/compose.css:
```diff
@@ -535,11 +535,49 @@ textarea.new_message_textarea,
+        margin: 1px;
+
+        &:hover {
+            background-color: hsl(231deg 100% 90% / 20%);
```
Is this a new redesign color? It might belong in a variable, if so (a quick grep doesn't show it elsewhere in the codebase)

shameondev replied:
It's just that the color was used in the demo. @terpimost, could you assist us here? Should we utilize a variable from the existing list? Or should we introduce a new variable? Here's a screenshot of the background we're discussing: 
![image](https://github.com/zulip/zulip/assets/53193850/819769db-137d-48be-a2c3-845ab385a522)

karlstolley commented:
I added a few inline queries here, Dan. I still need to poke at this a bit more in development, but this is looking good so far.

karlstolley commented:
Visually, this is looking good.

I'm not convinced that displaying the New Topic button only once any current topic is changed is the best choice, because the appearance of the button seems to suggest that it will _create_ the new topic, instead of clearing the box (and potentially frustrating a user who's spent time writing the perfect name for the topic stream).

It almost seems like we should go the other way: display the New Topic button in either an empty or unedited topic input box, and hide it when content is added or modified.

Thoughts?

karlstolley commented:
Once the New Topic button disappears, it doesn't reappear until the topic input is edited:

![disappearing-new-topic-button](https://github.com/zulip/zulip/assets/170719/4bb3f6b0-e03f-4088-aa7b-0c59770924ae)

shameondev commented:
> I'm not convinced that displaying the New Topic button only once any current topic is changed is the best choice, because the appearance of the button seems to suggest that it will create the new topic, instead of clearing the box (and potentially frustrating a user who's spent time writing the perfect name for the topic stream).

I agree with you here.

> It almost seems like we should go the other way: display the New Topic button in either an empty or unedited topic input box, and hide it when content is added or modified.


I'm not sure why we should display the "new topic" button if the input is empty. The main purpose of this button is to clear the input.

Should we discuss this on CZO? Or maybe we should stick to the original plan and test it on CZO to see how we feel about having this button that clears the input?

shameondev commented:
@karlstolley, I've updated the PR to fix the bug you discovered.

karlstolley commented:
> I'm not sure why we should display the "new topic" button if the input is empty. The main purpose of this button is to clear the input.
> 
> Should we discuss this on CZO? Or maybe we should stick to the original plan and test it on CZO to see how we feel about having this button that clears the input?

I think it's probably worth opening a discussion there, for sure.

shameondev commented:
@karlstolley, I've made updates to the PR and included screenshots reflecting the changes made to the compose box when it's in the closed state. I would greatly appreciate any feedback you have. Thank you!

shameondev commented:
I have also noticed that the "New topic" button is not properly aligned. I would appreciate any suggestions on how to fix this issue.
![image](https://github.com/zulip/zulip/assets/53193850/5b943e2a-6207-423a-b24f-96c241079586)

karlstolley commented:
Good eye. If you set `align-items: baseline` on `#compose_recipient_box`, it looks to me like the button will properly align _and_ the baseline of the input-box text and the New topic text will also properly align.

terpimost commented:
@karlstolley we shouldn't fanatically go for baseline alignment.
Because space between elements and their surroundings optically influence perception.
So font size in that button is smaller than the input and it make sense to align text inside the button vertically in the center

terpimost commented:
but again, I don't think we should be super nerdy in the old design implementation of that and we could revisit it in the new compose box design implementation

karlstolley commented:
Agreed. I don't want to be fanatical about anything :) but I also don't like littering the CSS up with little 1- or 2-pixel adjustments here and there if there's a more solid and flexible foundation we can build on (sometimes that's using baselines, other times that's using center alignment or plain old leaving the text alone and adjusting the padding of a button around it).

karlstolley commented:
The alignment looks to be OK for me this morning. My only lingering concern is "New topic" flashing up again after a user begins to compose a new topic. I expect that's going to be way too tempting for users to click after they've created a new topic.

Marking for integration review. I'll be eager to hear feedback once this is up on CZO; we should probably preemptively start a topic for it in #frontend or something as soon as it's live.

karlstolley commented:
Bumping this to priority integration review because it's a semi-blocker for the compose box work I'm prepping :)

timabbott commented:
I don't like how the styling of the button when not clicked looks identical to that of text; I don't have a specific solution but I think we need to address something there before merging.

![image](https://github.com/zulip/zulip/assets/2746074/0c4ae35a-f0a1-4bb3-a64a-5aa560f39941)

shameondev commented:
I've updated the PR with fixes for the incorrect width of the compose-box on mobile devices.

shameondev commented:
Hmm, I've noticed a regression in the styles after rebasing on the main branch. I think it might be due to this commit: [commit](https://github.com/zulip/zulip/commit/f717aa99a05b20c4ec4127616062a07709390bb6).

Hey @sahil839, what's the best approach to fix this? Should I simply add the styles I need in the appropriate place, right?

sahil839 commented:
Yes. We do not have any common close for the compose box input currently. You can refer to the CSS for `.modal_text_input` or `.settings_text_input` selectors as they do not have redundant CSS applied in bootstrap and those would probably be redundant here too.

shameondev commented:
I have resolved the merge conflicts and rebased the necessary changes after removing Bootstrap into the commit history. Additionally, I included the commit that modifies the colors of the "New topic" button and updated the screenshots.

It is now ready for review!

timabbott commented:
Updating the deployment on chat.zulip.org.

timabbott commented:
Updated this on chat.zulip.org.

shameondev commented:
The feedback that's left is in this [message](https://chat.zulip.org/#narrow/stream/137-feedback/topic/new.20topic.20button/near/1627213). Mostly, it's about renaming the buttons.

karlstolley commented:
Closing in favor of an extension/revision of this work at #27046. Thanks, Dan!
